### ISSUE TITLE
ProcessExitedException: process 0 terminated with signal SIGABRT

### ISSUE THREAD

**Issue by sadimanna** [pytorch/xla#3122]
I am trying to train a model on Google Colab TPUs. I installed pytorch-1.8.2(cpu) and pytorch_xla-1.8

When I run the demo notebook "PyTorch on Cloud TPUs: MultiCore Training AlexNet on Fashion MNIST", I am getting no error in the training process. But I am getting the following error when I run my code.

```
Process 0 is using TPU:0
Process 4 is using TPU:4
Process 5 is using TPU:5
Process 1 is using TPU:1
Process 7 is using TPU:7
Process 2 is using TPU:2
Process 6 is using TPU:6
Process 3 is using TPU:3
Exception in thread Thread-15:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/parallel_loader.py", line 166, in _worker
    batch = xm.send_cpu_data_to_device(batch, device)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 846, in send_cpu_data_to_device
    return ToXlaTensorArena(convert_fn, select_fn).transform(data)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 412, in transform
    self._convert()
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 384, in _convert
    self._converted_tensors = self._convert_fn(self._tensors)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 841, in convert_fn
    return torch_xla._XLAC._xla_tensors_from_aten(tensors, devices)
RuntimeError: tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:415 : Check failed: session->session()->Run(session_work->feed_inputs, session_work->outputs_handles, &outputs) == ::tensorflow::Status::OK() (Aborted: Session ab2f34ddd28f078d is not found. vs. OK)
*** Begin stack trace ***
	tensorflow::CurrentStackTrace()
	
	xla::util::MultiWait::Complete(std::function<void ()> const&)
	
	
	
	clone
*** End stack trace ***


Exception in device=TPU:0: Aborted: Session 3c850d7646b63255 is not found.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py", line 329, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py", line 323, in _start_fn
    fn(gindex, *args)
---------------------------------------------------------------------------
ProcessExitedException                    Traceback (most recent call last)
<ipython-input-7-b6cd7ba69c0e> in <module>()
----> 1 trainer.fit()

3 frames
/content/tpu_trainer.py in fit(self)
     85     def fit(self):
     86         torch.set_default_tensor_type('torch.FloatTensor')
---> 87         xmp.spawn(self.dev_fit, args=({},), nprocs = self.num_tpu_workers, start_method = 'fork')
     88 
     89     def dev_fit(self, index, flags):

/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    392         join=join,
    393         daemon=daemon,
--> 394         start_method=start_method)
    395 
    396 

/usr/local/lib/python3.7/dist-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    186 
    187     # Loop on join until it returns True or raises an exception.
--> 188     while not context.join():
    189         pass
    190 

/usr/local/lib/python3.7/dist-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    134                     error_pid=failed_process.pid,
    135                     exit_code=exitcode,
--> 136                     signal_name=name
    137                 )
    138             else:

ProcessExitedException: process 0 terminated with signal SIGABRT
```

I only instantiate my model once using `xmp.MpModelWrapper` and within the `_map_fn`, I do like this 

`new_model = WRAPPERMODEL.to(device)`

What is the reason for this kind of error?? OR what causes these kinds of errors??

**Comment by sadimanna** [comment]
Also receiving "ProcessExitedException: process 0 terminated with exit code 17"

```
Process 0 is using TPU:0
Process 2 is using TPU:2
Epoch 1 train begin 16:10:28
Process 7 is using TPU:7
Process 4 is using TPU:4
Process 1 is using TPU:1
Process 3 is using TPU:3
Process 6 is using TPU:6
Process 5 is using TPU:5
Exception in device=TPU:0: Aborted: Session 1a625b17752b2794 is not found.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py", line 329, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py", line 323, in _start_fn
    fn(gindex, *args)
  File "<ipython-input-9-dbc82064644c>", line 87, in _mp_fn
    train_mnist(index, {})
  File "<ipython-input-9-dbc82064644c>", line 71, in train_mnist
    total_loss = train_loop_fn(train_device_loader)
  File "<ipython-input-9-dbc82064644c>", line 48, in train_loop_fn
    total_loss += loss.item()
RuntimeError: Aborted: Session 1a625b17752b2794 is not found.
---------------------------------------------------------------------------
ProcessExitedException                    Traceback (most recent call last)
<ipython-input-9-dbc82064644c> in <module>()
     89 
     90 if __name__ == '__main__':
---> 91     xmp.spawn(_mp_fn, args=({},), nprocs=8, start_method = 'fork')

2 frames
/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py in spawn(fn, args, nprocs, join, daemon, start_method)
    392         join=join,
    393         daemon=daemon,
--> 394         start_method=start_method)
    395 
    396 

/usr/local/lib/python3.7/dist-packages/torch/multiprocessing/spawn.py in start_processes(fn, args, nprocs, join, daemon, start_method)
    186 
    187     # Loop on join until it returns True or raises an exception.
--> 188     while not context.join():
    189         pass
    190 

/usr/local/lib/python3.7/dist-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    142                     error_index=error_index,
    143                     error_pid=failed_process.pid,
--> 144                     exit_code=exitcode
    145                 )
    146 

ProcessExitedException: process 0 terminated with exit code 17
```

**Comment by sadimanna** [comment]
Another such exception is occurring when using `loss.cpu().item()`

```
Process 0 is using TPU:0
Process 6 is using TPU:6
Epoch 1 train begin 16:34:47
Process 3 is using TPU:3
Process 5 is using TPU:5
Process 7 is using TPU:7
Process 4 is using TPU:4
Process 2 is using TPU:2
Process 1 is using TPU:1
Exception in thread Thread-12:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/parallel_loader.py", line 166, in _worker
    batch = xm.send_cpu_data_to_device(batch, device)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 846, in send_cpu_data_to_device
    return ToXlaTensorArena(convert_fn, select_fn).transform(data)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 412, in transform
    self._convert()
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 384, in _convert
    self._converted_tensors = self._convert_fn(self._tensors)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/core/xla_model.py", line 841, in convert_fn
    return torch_xla._XLAC._xla_tensors_from_aten(tensors, devices)
RuntimeError: tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:415 : Check failed: session->session()->Run(session_work->feed_inputs, session_work->outputs_handles, &outputs) == ::tensorflow::Status::OK() (Aborted: Session ab7644a8c1cef902 is not found. vs. OK)
*** Begin stack trace ***
	tensorflow::CurrentStackTrace()
	
	xla::util::MultiWait::Complete(std::function<void ()> const&)
	
	
	
	clone
*** End stack trace ***


Exception in device=TPU:6: Aborted: Session a720a2478eb37ca3 is not found.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py", line 329, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/usr/local/lib/python3.7/dist-packages/torch_xla/distributed/xla_multiprocessing.py", line 323, in _start_fn
    fn(gindex, *args)
  File "<ipython-input-9-e74a784fd62d>", line 87, in _mp_fn
    train_mnist(index, {})
  File "<ipython-input-9-e74a784fd62d>", line 71, in train_mnist
    total_loss = train_loop_fn(train_device_loader)
  File "<ipython-input-9-e74a784fd62d>", line 48, in train_loop_fn
    total_loss += loss.cpu().item()
RuntimeError: Aborted: Session a720a2478eb37ca3 is not found.
---------------------------------------------------------------------------
ProcessExitedException                    Traceback (most recent call last)
<ipython-input-9-e74a784fd62d> in <module>()
     89 
     90 if __name__ == '__main__':
---> 91     xmp.spawn(_mp_fn, args=({},), nprocs=8, start_method = 'fork')

2 frames
/usr/local/lib/python3.7/dist-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    142                     error_index=error_index,
    143                     error_pid=failed_process.pid,
--> 144                     exit_code=exitcode
    145                 )
    146 

ProcessExitedException: process 6 terminated with exit code 17
```

**Comment by JackCaoG** [comment]
Thanks for reporting.
```
RuntimeError: tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:415 : Check failed: session->session()->Run(session_work->feed_inputs, session_work->outputs_handles, &outputs) == ::tensorflow::Status::OK() (Aborted: Session ab7644a8c1cef902 is not found. vs. OK)
```
is just a symptom and it usually means remote server crashed. I will try to repo your code, by running `MultiCore Training AlexNet on Fashion MNIST` myself.

**Comment by sadimanna** [comment]
@JackCaoG , I am generally not getting any error when running `MultiCore Training AlexNet on Fashion MNIST` as I have mentioned. I am getting the error when I am running my code with custom loss functions and custom DataSet modules. 

However, I would like to know if this is caused by memory allocation issue? Like, if the model is large or the batch size is too big?

**Comment by sadimanna** [comment]
@JackCaoG 

I have compiled my code in this Google Colab notebook! 
I am pasting the link here. I hope it will help getting some more info about the issue!!

https://colab.research.google.com/drive/1Z58w4DhNgxLVITRz2A-j-OoasZGXjE-e?usp=sharing

**Comment by JackCaoG** [comment]
Thanks, I will try to repo it in next few days

**Comment by miladm** [comment]
Thanks for filing this issue! I will look into this in a next few days.

**Comment by sadimanna** [comment]
> @JackCaoG
> 
> I have compiled my code in this Google Colab notebook!
> I am pasting the link here. I hope it will help getting some more info about the issue!!
> 
> https://colab.research.google.com/drive/1Z58w4DhNgxLVITRz2A-j-OoasZGXjE-e?usp=sharing

@miladm Please check this code! This is my code on Google Colab where I was getting the error! This may help !

**Comment by miladm** [comment]
@sadimanna thanks for the link. I am looking into reproducing it.

**Comment by miladm** [comment]
@sadimanna 

+ I was able to internally run your code and see the underlying error. So far, it seems this issue is related to exceeding the memory space - WIP.

+ Can you please point me to the lines of code (loss functions, DataSet modules) you customized? Also do you make any changes to the input data? I assume not.

+ I spent some time adjusting your model parameters, and noticed when reducing `BTnet.proj_num_layers = 1` your model continues to train without the initial failure. I let it train for `~10min` and it didn't see a failure; feel free to take a deeper look and let me know what you discover.

+ RE: `"However, I would like to know if this is caused by memory allocation issue? Like, if the model is large or the batch size is too big?"` <- Feel free to try adjusting your model / batch size to see if the issue is due to data vs. model.

**Comment by sadimanna** [comment]
@miladm 

* The loss functions have been coded as per this https://github.com/facebookresearch/barlowtwins, official PyTorch implementation of the algorithm.

* I made the CIFAR10DataModule in the given notebook to download the data and return a dataloader which gives a batch of size B.

* I did not make any changes to the input data. I just download the data and stored it in an array.

* I will run the code like you suggested and get back to you ASAP!!
